core/src/main/java/org/apache/accumulo/core/clientImpl/TabletServerBatchWriter.java

```diff
@@ -1028,14 +1020,139 @@ private MutationSet sendMutationsToTabletServer(String location,
         timeoutTracker.errorOccured();
         throw new IOException(e);
       } catch (TApplicationException tae) {
+        // no need to bother closing session in this case
+        usid = null;
         updateServerErrors(location, tae);
         throw new AccumuloServerException(location, tae);
       } catch (ThriftSecurityException e) {
+        // no need to bother closing session in this case
+        usid = null;
         updateAuthorizationFailures(
             tabMuts.keySet().stream().collect(toMap(identity(), ke -> e.code)));
         throw new AccumuloSecurityException(e.user, e.code, e);
       } catch (TException e) {
         throw new IOException(e);
+      } finally {
+        if (usid != null) {
+          // There is an open session, must close it before the batchwriter closes or writes could
+          // happen after the batch writer closes. See #3721. Queuing a task instead of executing
+          // the code here because throwing exceptions in a finally block makes the code hard to
+          // reason about. Queue is less likely to throw an exception.
+          //
+          // When the task is created below it adds to the failedSessions map, this is done while
+          // the mutations for this task are still incremented in totalMemUsed. It is important that
+          // these overlap in time so that there is no gap where the client would not wait for the
+          // session to close. Overlap in times means the session is added to failedSessions before
+          // the mutations are decremented from totalMemUsed.
+          sendThreadPool.execute(new CloseSessionTask(location, usid));
+        }
+      }
+    }
+
+    class CloseSessionTask implements Runnable {
```

Comment:
There's a lot going on in this. This type should be in its own file. If you want it to have access to TabletServerBatchWriter internals, you can pass `this` as a parameter.

Reply:
The class is very tightly coupled with the internal implementation and accesses private functions and instance variable of the containing class.  So to push it out would also require changing things to be non private.   I do not think that internal classes that are very tightly coupled with internal  implementation of the containing class need to be moved out.   Its only a class because its internal code that needs to be run in a another thread.

Reply:
I think that the tight coupling might be the problem. It strains the principle of encapsulation by making everything accessible to everything else within a very large class. I think a more modular approach would make this very large class a bit more comprehensible by forcing clear APIs. We've been in this situation before, and it has benefited us to separate things out. I don't think this is that different. I'm not suggesting it all be done here... I'm just suggesting the new class doesn't need to add to the size and complexity of the already large containing class.

Reply:
I think the tight coupling is nice when analyzing the code. I like when I can examine references to a batch writer variable in an IDE and see all the functions that directly use it. Makes it easy to analyze the code for correctness w/ multithreading.  In  other parts of the Accumulo code we have layers of evolved indirection in the code that make it much more difficult to analyze, understand, and maintain.  I think this code could benefit from being made more modular, but making it more modular should be done with an overall strategy in mind.  I think a strategy of simply extracting inner classes does not evolve to any particular goal and can make the code harder to understand and maintain when iterating that strategy. Maybe a plan like separating  the RPC code from the in in memory state of the batch writer would be a better way to achieve a modular batch writer, not sure.